File: Backend/src/main/java/com/example/medisphere/service/AppointmentService.java
package com.example.medisphere.service;

import com.example.medisphere.model.Appointment;
import com.example.medisphere.repository.AppointmentRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class AppointmentService {

    private final AppointmentRepository appointmentRepository;

    public AppointmentService(
            AppointmentRepository appointmentRepository) {

        this.appointmentRepository = appointmentRepository;
    }

    // Create appointment
    public Appointment createAppointment(
            Appointment appointment) {

        return appointmentRepository.save(appointment);
    }

    // Get all appointments
    public List<Appointment> getAllAppointments() {

        return appointmentRepository.findAll();
    }

    // Get appointments by patient ID
    public List<Appointment> getAppointmentsByPatientId(
            String patientId) {

        return appointmentRepository.findByPatientId(
                patientId
        );
    }

    // Get appointment by ID
    public Optional<Appointment> getAppointmentById(
            String id) {

        return appointmentRepository.findById(id);
    }

    // Update appointment
    public Appointment updateAppointment(
            String id,
            Appointment updatedAppointment) {

        return appointmentRepository.findById(id)
                .map(appointment -> {

                    appointment.setPatientId(
                            updatedAppointment.getPatientId()
                    );

                    appointment.setDoctorId(
                            updatedAppointment.getDoctorId()
                    );

                    appointment.setAppointmentDate(
                            updatedAppointment.getAppointmentDate()
                    );

                    appointment.setAppointmentTime(
                            updatedAppointment.getAppointmentTime()
                    );

                    appointment.setStatus(
                            updatedAppointment.getStatus()
                    );

                    appointment.setReason(
                            updatedAppointment.getReason()
                    );

                    return appointmentRepository.save(
                            appointment
                    );
                })
                .orElse(null);
    }

    // Delete appointment
    public boolean deleteAppointment(String id) {

        if (appointmentRepository.existsById(id)) {

            appointmentRepository.deleteById(id);

            return true;
        }

        return false;
    }
}

File: Frontend/src/services/appointmentService.js
const API_URL = "http://localhost:8080/api/appointments";

// Get all appointments
export const getAppointments = async () => {
  const response = await fetch(API_URL);

  if (!response.ok) {
    throw new Error("Failed to fetch appointments");
  }

  return response.json();
};

// Get appointments by patient ID
export const getAppointmentsByPatientId = async (patientId) => {
  const response = await fetch(`${API_URL}/patient/${patientId}`);

  if (!response.ok) {
    throw new Error("Failed to fetch patient appointments");
  }

  return response.json();
};

// Get appointment by ID
export const getAppointmentById = async (id) => {
  const response = await fetch(`${API_URL}/${id}`);

  if (!response.ok) {
    throw new Error("Failed to fetch appointment");
  }

  return response.json();
};

// Create appointment
export const createAppointment = async (appointment) => {
  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(appointment),
  });

  if (!response.ok) {
    throw new Error("Failed to create appointment");
  }

  return response.json();
};

// Update appointment
export const updateAppointment = async (id, appointment) => {
  const response = await fetch(`${API_URL}/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(appointment),
  });

  if (!response.ok) {
    throw new Error("Failed to update appointment");
  }

  return response.json();
};

// Delete appointment
export const deleteAppointment = async (id) => {
  const response = await fetch(`${API_URL}/${id}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error("Failed to delete appointment");
  }

  return true;
};